content/en/docs/unit-testing.md

```diff
@@ -0,0 +1,197 @@
+function. `s3GetObjectAPI` is satisfied by the {{% alias service=S3 %}}
+client's [GetObject]({{< apiref "service/s3#Client.GetObject" >}}) method.
+
+```go
+import "context"
+import "github.com/aws/aws-sdk-go-v2/service/s3"
+
+// ...
+
+type S3GetObjectAPI interface {
+	PutObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error)
+}
+
+func GetObjectFromS3(api S3GetObjectAPI, bucket, key string) ([]byte, error) {
+	object, err := api.PutObject(context.TODO(), &s3.GetObjectInput{
+		Bucket: &bucket,
+		Key:    &key,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	all, err := ioutil.ReadAll(object.Body)
+	if err != nil {
+		return nil, err
+	}
+	return all, nil
+}
+```
+
+To test the `GetObjectFromS3` function, `mockGetObjectAPI` is used to satisfy
+the `s3GetObjectAPI` interface definition. The mock type is then to mock output
+and error responses returned from the service client.
+
+```go
+import "testing"
+import "github.com/aws/aws-sdk-go-v2/service/s3"
+
+// ...
+
+type mockGetObjectAPI func(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error)
+
+func (m mockGetObjectAPI) PutObject(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error) {
+	return m(ctx, params, optFns...)
+}
+
+func TestGetObjectFromS3(t *testing.T) {
+	cases := []struct {
+		client func(t *testing.T) s3GetObjectAPI
+		bucket string
+		key    string
+		expect []byte
+	}{
+		{
+			client: func(t *testing.T) s3GetObjectAPI {
+				return mockGetObjectAPI(func(ctx context.Context, params *s3.GetObjectInput, optFns ...func(*s3.Options)) (*s3.GetObjectOutput, error) {
+					t.Helper()
+					if params.Bucket == nil {
+						t.Fatal("expect bucket to not be nil")
+					}
+					if e, a := "fooBucket", *params.Bucket; e != a {
+						t.Errorf("expect %v, got %v", e, a)
+					}
+					if params.Key == nil {
+						t.Fatal("expect key to not be nil")
+					}
+					if e, a := "barKey", *params.Key; e != a {
+						t.Errorf("expect %v, got %v", e, a)
+					}
+
+					return &s3.GetObjectOutput{
+						Body: ioutil.NopCloser(bytes.NewReader([]byte("this is the body foo bar baz"))),
+					}, nil
+				})
+			},
+			bucket: "fooBucket",
+			key:    "barKey",
+			expect: []byte("this is the body foo bar baz"),
+		},
+	}
+
+	for i, tt := range cases {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			content, err := GetObjectFromS3(tt.client(t), tt.bucket, tt.key)
+			if err != nil {
+				t.Fatalf("expect no error, got %v", err)
+			}
+			if e, a := tt.expect, content; bytes.Compare(e, a) != 0 {
+				t.Errorf("expect %v, got %v", e, a)
+			}
+		})
+	}
+}
+```
+
+## Mocking Paginators
+
+In this example, `ListObjectsV2Pager` is an interface that defines the
+behaviors for the {{% alias service=S3 %}}
+[ListObjectsV2Paginator]({{< apiref "service/s3#ListObjectsV2Paginator" >}}).
+required by `CountObjects` function.
+
```

Comment:
Should this call out that the user could mock out the API client returning pages of response?  That may be easier to mock out than the paginator, since the user would need to mock out the internal behavior of the paginator to be similar to the SDK's implementation.

Reply:
Hmmm, this is maybe an equally confusing situation. As mocking it in that way also requires the user to understand the population of NextToken and the concept of not returning it for the last response in their list. Arguably mocking at the client level requires more knowledge about the API and how pagination works then the actual usage of the paginator.